docs/markdown/Writing Plugins/tutorials/01-create-new-goal-with-a-custom-target.md

```diff
@@ -0,0 +1,335 @@
+    "dependencies": [],
+    "description": null,
+    "source_raw": "VERSION",
+    "sources": [
+      "myapp/VERSION"
+    ],
+    "tags": null
+  }
+]
+```
+
+## Creating a goal
+
+[Goals](https://www.pantsbuild.org/docs/goals) are the commands that Pants runs such as `fmt` or `lint`. To be able to get a project version number (using a target we've just created), we need to [create a new goal](https://www.pantsbuild.org/docs/rules-api-goal-rules). The code below is the boilerplate necessary to create a goal, so it's not really necessary to understand how, for instance, subsystems work right now. The function decorated with the `@goal_rule` can be named anything, but it's helpful for the name to represent the functionality your goal provides. To make your goal part of the plugin's interface, add it to the `rules` function in the `register.py` module.
+
+```python pants-plugins/project_version/rules.py
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import collect_rules, goal_rule
+
+
+class ProjectVersionSubsystem(GoalSubsystem):
+    name = "project-version"
+    help = "Show representation of the project version from the `VERSION` file."
+
+
+class ProjectVersionGoal(Goal):
+    subsystem_cls = ProjectVersionSubsystem
+
+
+@goal_rule
+async def goal_show_project_version() -> ProjectVersionGoal:
+    return ProjectVersionGoal(exit_code=0)
+
+
+def rules():
+    return collect_rules()
+```
+```python pants-plugins/project_version/rules.py
+from project_version.targets import ProjectVersionTarget
+import project_version.rules as project_version_rules
+
+
+def target_types():
+    return [ProjectVersionTarget]
+
+
+def rules():
+    return [*project_version_rules.rules()]
+```
+
+You can now run `./pants project-version` to confirm the command exits with the exit code `0`.
+
+At this point, we are ready to do something useful with the new target of ours. Goals generally run on targets, so they need to be passed as an argument in the command line. For instance, to format the `myproject` directory targets, you would run `./pants fmt myproject`. To get the version of a project in your repository, it makes sense to pass to the `project-version` goal a project directory containing the `version_file` definition.
+
+To make a target passed as an argument accessible in the goal rule, we pass the [`Targets`](https://www.pantsbuild.org/docs/rules-api-goal-rules#how-to-operate-on-targets) as input arguments of the function along with the [`Console`](https://www.pantsbuild.org/docs/rules-api-goal-rules#console-output-to-stdoutstderr) object so that we can print the details of our target in the user terminal:
+
+```python
+@goal_rule
+async def goal_show_project_version(console: Console, targets: Targets) -> ProjectVersionGoal:
+    # since we don't know what targets will be passed (e.g. `::`), we want to keep only `version_file` targets
+    targets = [tgt for tgt in targets if tgt.alias == ProjectVersionTarget.alias]
```

Comment:
And here we can circle back to using a dedicated source field again, as it'll be more stable/flexible to filter the targets based on the fact that they have a `ProjectVersionSourceField`, rather than just a single targets alias.

More stable, in the notion that if you down the line deprecate the target alias in order to rename it, this code does not handle both the deprecated and new alias at the same time.
More flexible, as other plugins may import your `ProjectVersionSourceField` and use it in their own custom targets, and the project-version stuff here would still apply to those targets as well :)

Reply:
Thank you, Andreas! I am going to include this in my second tutorial, if that's okay!